README.md

```diff
@@ -229,6 +243,253 @@ Benchee.run %{"Awesome stuff" => fn -> magic end }
+Local hooks are defined as part of the definition of the benchmarking function. Therefore in the initial map instead of just a single function a tuple can be provided where the bencharking function comes first and then a list of keyword options follows specifying the hooks to run:
+
+```elixr
+Benchee.run %{
+  "foo" => {fn(input) -> ... end, before_scenario: fn(input) -> ... end},
+  "bar" => fn(input) -> ... end
+}, inputs: %{
+  "input 1" => 1,
+  "input 2" => 2
+}
+```
+
+Here `before_scenario` is only run for the 2 scenarios associated with `"foo"`, i.e. foo with input 1 and foo with input 2. It is _not_ run for any `"bar"` benchmarks.
+
+#### before_scenario
+
+Is executed before every [scenario](#what-is-a-scenario) that it applies to considering global/local configuration. `before_scenario` takes the input of the scenario as an argument.
+
+If you haven't defined `inputs`, it is still called with an argument albeit the special input which is returned by `Benchee.Benchmark.no_input()`. This also holds true for `before_each` and `after_scenario`.
+
+The defined function **must return whatever it wants following hooks and the benchmarking function to consider the input**. This means, that there usually are 3 kinds of before_scenarios:
+
+* you just want to invoke a side effect, in that case return the given input unchanged
+* you want to alter the given input, in that case alter the given input
+* you want to keep the given input but add some other data, in that case just return a tuple like `{original_input, new_fancy_data}`
+
+First the _global_ `before_scenario` is invoked, then the local one. This also means that the return value of the _global_ `before_scenario` is passed on into its _local_ sibling. That return value is then passed into `before_each` (if present) or the benchmarking function (if no `before_each` is present).
+
+Usage:
+
+```elixir
+Benchee.run %{
+  "foo" =>
+    {
+      fn(input) -> foo(input) end, # input is 3 here
+      before_scenario: fn(input) -> input + 1 end # input is 2 here
+    },
+  "bar" =>
+    fn(input) -> bar(input) end # input is 2 here (only global ran)
+}, inputs: %{
+  "input 1" => 1,
+  before_scenario: fn(input) -> input + 1 end # input is 1 here
+}
+```
+
+When might this be useful?
+* You want to start a process to be used in your scenario
+* Something that you want to use in the benchmark starts a process associated with the PID of `self()` (each scenario is executed in its own process so these can't be started before all)
+* Clearing the cache before a scenario
+
+#### before_each
+
+Is executed before each invocation of the benchmarking function/before every measurement. The input for `before_each` is either the `input` or what was returned by `before_scenario`. It works much like [`before_scenario`](#before_scenario), the only major difference being that it is invoked at another time. Most importantly, the function needs to return what the benchmarking function should receive as input.
+
```

Comment:
This section basically repeats what was already explained in`before_scenario` concerning inputs and return values. To me, reiterating that information in another spot makes it seem more complicated than it actually is :)

Reply:
Huh, that's an interesting point.

For me I reiterated it to underline the point and make people not miss it. Sadly my recent experience maintaining simplecov has taught me that people often really just skim the README - and don't properly read not even before opening an issue.

So, wanted to make that extra clear :)

Reply:
Not sure what the best way is here, any idea?